File: src/core/ptz_incremental_optimizer.cc
/*
 * Author: mingjian (lory.gjh@alibaba-inc.com)
 * Created Date: 2025-02-17 19:53:41
 * Modified By: mingjian (lory.gjh@alibaba-inc.com)
 * Last Modified: 2025-02-18 14:09:51
 * -----
 * Copyright (c) 2025 Alibaba Inc.
 */

#include "ptz_incremental_optimizer.h"

#include <iostream>
#include <limits>

#include "data_io.h"
#include "krt_optimizer.h"
#include "ptzray_optimizer.h"
#include "utils/logging.h"

using namespace std;

namespace ptzcalib {

namespace {

void SortIndicesByConfidence(vector<long>& indices, const vector<float>& confidences_rank)
{
  sort(indices.begin(), indices.end(), [&](long A, long B) -> bool {
    if (confidences_rank[A] == confidences_rank[B])
      return A < B;
    return confidences_rank[A] > confidences_rank[B];
  });
}

}  // namespace

long PtzIncrementalOptimizer::kMaxNumImages = 100000;
float PtzIncrementalOptimizer::kBaGlobalImagesRatio = 1.1;

PtzIncrementalOptimizer::PtzIncrementalOptimizer(const vector<ImageFeatures>& features, const vector<MatchesInfo>& matches_info,
                                                 const vector<Camera>& cameras, int max_iter)
    : cameras_(cameras), features_(features), matches_info_(matches_info), max_iter_(max_iter)
{
}

PtzIncrementalOptimizer::PtzIncrementalOptimizer(const vector<ImageFeatures>& features, const vector<MatchesInfo>& matches_info,
                                                 const vector<Camera>& cameras, const std::vector<std::string>& names, int max_iter)
    : cameras_(cameras), features_(features), matches_info_(matches_info), names_(names), max_iter_(max_iter)
{
}

bool PtzIncrementalOptimizer::Solve(vector<Camera>& cameras, std::unordered_set<long>& reg_image_ids)
{
  if (!CheckValid())
    return false;

  const int kInitNumTrials = 50;
  for (int num_trials = 0; num_trials < kInitNumTrials; ++num_trials) {
    // Register initial pair
    long image_id1, image_id2;
    bool find_init_success = FindInitialImagePair(image_id1, image_id2);
    if (!find_init_success) {
      PLOGI << "No good initial image pair found";
      return false;
    }

    PLOGI << "Initializing with image pair #" << image_id1 << " and #" << image_id2;

    bool reg_init_success = RegisterInitialImagePair(image_id1, image_id2);
    if (!reg_init_success) {
      PLOGI << "Initialization failed - possible solutions:" << endl
            << "     - try to relax the initialization constraints" << endl
            << "     - manually select an initial image pair";
      continue;
    }
    else {
      PLOGI << "Initialization success. Focal: " << cameras_[image_id1].K().at<double>(0, 0) << " , "
            << cameras_[image_id2].K().at<double>(0, 0);
    }

    AdjustGlobalBundle();

    size_t ba_prev_num_reg_images = NumRegImages();

    bool reg_next_success = true;
    while (reg_next_success) {
      reg_next_success = false;

      const auto next_image_ids = FindNextImages();

      if (next_image_ids.empty()) {
        break;
      }

      for (size_t reg_trial = 0; reg_trial < next_image_ids.size(); ++reg_trial) {
        const auto image_id = next_image_ids[reg_trial];
        reg_next_success = RegisterNextImage(image_id);

        PLOGI << "Register image #" << image_id << (reg_next_success ? " success" : " failed")
              << ", focal: " << cameras_[image_id].K().at<double>(0, 0) << ". Currently registered: " << NumRegImages()
              << ", total: " << features_.size();

        if (reg_next_success) {
          if (NumRegImages() >= kBaGlobalImagesRatio * ba_prev_num_reg_images) {
            bool gba_success = AdjustGlobalBundle();

            if (gba_success) {
              ba_prev_num_reg_images = NumRegImages();
              break;
            }
            else {
              reg_image_ids_.erase(image_id);  // TODO: erase all cameras which not pass global ba
              reg_next_success = false;
            }
          }
        }

        if (!reg_next_success) {
          PLOGI << "Could not register, trying another image";

          // If initial pair fails to continue for some time,
          // abort and try different initial pair.
          const long kMinNumInitialRegTrials = 30;
          const int kMinModelSize = 3;
          if (reg_trial >= kMinNumInitialRegTrials && NumRegImages() < static_cast<size_t>(kMinModelSize)) {
            break;
          }
        }
      }
    }

    AdjustGlobalBundle();

    reg_image_ids = reg_image_ids_;
    cameras = cameras_;

    return true;
  }
}

void PtzIncrementalOptimizer::SetSeedImageId(const std::vector<long>& image_ids)
{
  PLOGI << "Manually set seed image ids";
  seed_image_ids_ = image_ids;
}

bool PtzIncrementalOptimizer::CheckValid()
{
  if (features_.empty() || features_.size() != cameras_.size() || max_iter_ <= 0)
    return false;
  else
    return true;
}

bool PtzIncrementalOptimizer::FindInitialImagePair(long& image_id1, long& image_id2)
{
  vector<long> image_ids1;
  if (seed_image_ids_.empty()) {
    image_ids1 = FindFirstInitialImage();
  }
  else {
    image_ids1 = seed_image_ids_;
  }

  for (long i1 = 0; i1 < image_ids1.size(); ++i1) {
    image_id1 = image_ids1[i1];

    const vector<long> image_ids2 = FindSecondInitialImage(image_id1);

    for (long i2 = 0; i2 < image_ids2.size(); ++i2) {
      image_id2 = image_ids2[i2];

      const long pair_id = ImagePairToPairId(image_id1, image_id2);

      // Try every pair only once.
      if (init_image_pairs_.count(pair_id) > 0) {
        continue;
      }

      init_image_pairs_.insert(pair_id);

      return true;
    }
  }

  image_id1 = numeric_limits<long>::max();
  image_id2 = numeric_limits<long>::max();
  return false;
}

vector<long> PtzIncrementalOptimizer::FindFirstInitialImage() const
{
  size_t num_imgs = features_.size();
  vector<float> confidences_rank(num_imgs, 0.0f);

  for (const auto& match_info : matches_info_) {
    long src_idx = match_info.src_img_idx;
    long dst_idx = match_info.dst_img_idx;
    float confidence = match_info.confidence;

    confidences_rank[src_idx] += confidence;
    confidences_rank[dst_idx] += confidence;
  }

  vector<long> indices(num_imgs);
  iota(indices.begin(), indices.end(), 0);
  SortIndicesByConfidence(indices, confidences_rank);

  vector<long> sorted_images_ids;
  for (auto& index : indices) {
    if (confidences_rank[index] <= 0.0f)
      break;
    sorted_images_ids.push_back(index);
  }

  return sorted_images_ids;
}

vector<long> PtzIncrementalOptimizer::FindSecondInitialImage(long image_id1) const
{
  size_t num_imgs = features_.size();
  vector<float> confidences_rank(num_imgs, 0.0f);

  for (const auto& match_info : matches_info_) {
    long src_idx = match_info.src_img_idx;
    long dst_idx = match_info.dst_img_idx;
    float confidence = match_info.confidence;
    const auto& matches = match_info.matches;

    const float kMinPixelDiff = 50;
    if (matches.empty())
      continue;
    if (image_id1 != src_idx && image_id1 != dst_idx)
      continue;
    if (image_id1 == src_idx && image_id1 == dst_idx)
      continue;
    if (CalPixelDiff(src_idx, dst_idx, matches) < kMinPixelDiff)
      continue;
    if (image_id1 == src_idx && image_id1 != dst_idx)
      confidences_rank[dst_idx] += confidence;
    if (image_id1 != src_idx && image_id1 == dst_idx)
      confidences_rank[src_idx] += confidence;
  }

  vector<long> indices(num_imgs);
  iota(indices.begin(), indices.end(), 0);
  SortIndicesByConfidence(indices, confidences_rank);

  vector<long> sorted_images_ids;
  for (auto& index : indices) {
    if (confidences_rank[index] <= 0.0f)
      break;
    sorted_images_ids.push_back(index);
  }

  return sorted_images_ids;
}

std::vector<long> PtzIncrementalOptimizer::FindNextImages() const
{
  size_t num_imgs = features_.size();
  vector<float> confidences_rank(num_imgs, 0.0f);

  for (const auto& match_info : matches_info_) {
    long src_idx = match_info.src_img_idx;
    long dst_idx = match_info.dst_img_idx;
    float confidence = match_info.confidence;

    if (src_idx == dst_idx)
      continue;
    if (match_info.H.empty())
      continue;

    // Only try registration for a certain maximum number of times.
    const size_t kMaxRegTrials = 4;
    if (num_reg_trials_.count(src_idx) != 0 && num_reg_trials_.at(src_idx) > kMaxRegTrials)
      continue;
    if (num_reg_trials_.count(dst_idx) != 0 && num_reg_trials_.at(dst_idx) > kMaxRegTrials)
      continue;

    if (reg_image_ids_.find(src_idx) != reg_image_ids_.end() && reg_image_ids_.find(dst_idx) != reg_image_ids_.end()) {
      // Skip images that are already registered.
      continue;
    }
    else if (reg_image_ids_.find(src_idx) == reg_image_ids_.end() && reg_image_ids_.find(dst_idx) == reg_image_ids_.end()) {
      // Skip images that are not neighbor.
      continue;
    }
    else if (reg_image_ids_.find(src_idx) != reg_image_ids_.end() && reg_image_ids_.find(dst_idx) == reg_image_ids_.end()) {
      confidences_rank[dst_idx] += confidence;
    }
    else if (reg_image_ids_.find(src_idx) == reg_image_ids_.end() && reg_image_ids_.find(dst_idx) != reg_image_ids_.end()) {
      confidences_rank[src_idx] += confidence;
    }
  }

  vector<long> indices(num_imgs);
  iota(indices.begin(), indices.end(), 0);
  SortIndicesByConfidence(indices, confidences_rank);

  vector<long> sorted_images_ids;
  for (auto& index : indices) {
    if (confidences_rank[index] <= 0.0f)
      break;
    sorted_images_ids.push_back(index);
  }

  return sorted_images_ids;
}

float PtzIncrementalOptimizer::CalPixelDiff(long image_id1, long image_id2, const vector<cv::DMatch>& matches) const
{
  size_t num_matches = matches.size();
  float total_dists = 0.0f;

  for (size_t i = 0; i < num_matches; ++i) {
    const cv::DMatch& m = matches[i];
    auto pt1 = features_[image_id1].keypoints[m.queryIdx].pt;
    auto pt2 = features_[image_id2].keypoints[m.trainIdx].pt;

    total_dists += cv::norm(pt1 - pt2);
  }

  return total_dists * 1.0f / num_matches;
}

long PtzIncrementalOptimizer::ImagePairToPairId(long image_id1, long image_id2) const
{
  if (image_id1 < image_id2)
    return image_id1 * kMaxNumImages + image_id2;
  else
    return image_id2 * kMaxNumImages + image_id1;
}

void PtzIncrementalOptimizer::SetInitialImagePairParameters(long image_id1, long image_id2)
{
  constexpr double ratio = 1.2;  // this affects fov
  double focal = ratio * max(features_[image_id1].img_size.width, features_[image_id1].img_size.height);
  double cx = 0.5 * features_[image_id1].img_size.width;
  double cy = 0.5 * features_[image_id1].img_size.height;
  cameras_[image_id1].K().at<double>(0, 0) = cameras_[image_id1].K().at<double>(1, 1) = focal;
  cameras_[image_id1].K().at<double>(0, 2) = cx;
  cameras_[image_id1].K().at<double>(1, 2) = cy;
  cameras_[image_id1].R() = cv::Mat_<double>::eye(3, 3);

  focal = ratio * max(features_[image_id2].img_size.width, features_[image_id2].img_size.height);
  cx = 0.5 * features_[image_id2].img_size.width;
  cy = 0.5 * features_[image_id2].img_size.height;
  cameras_[image_id2].K().at<double>(0, 0) = cameras_[image_id2].K().at<double>(1, 1) = focal;
  cameras_[image_id2].K().at<double>(0, 2) = cx;
  cameras_[image_id2].K().at<double>(1, 2) = cy;

  for (const auto& match_info : matches_info_) {
    long i = match_info.src_img_idx;
    long j = match_info.dst_img_idx;

    if (i == image_id1 && j == image_id2) {
      cv::Mat H_j_i = match_info.H;
      cv::Mat R_j_i = cameras_[j].K().inv() * H_j_i * cameras_[i].K();

      cameras_[j].R() = R_j_i * cameras_[i].R();
      break;
    }
  }
}

bool PtzIncrementalOptimizer::RegisterInitialImagePair(long image_id1, long image_id2)
{
  assert(NumRegImages() == 0);

  num_reg_trials_[image_id1] += 1;
  num_reg_trials_[image_id2] += 1;

  long pair_id = ImagePairToPairId(image_id1, image_id2);
  init_image_pairs_.insert(pair_id);

  SetInitialImagePairParameters(image_id1, image_id2);

  PTZRayOptimizer optimizer(features_, matches_info_, cameras_, {image_id1, image_id2}, max_iter_, PTZRay);
  bool ba_success = optimizer.Solve(cameras_);

  if (ba_success) {
    reg_image_ids_.insert(image_id1);
    reg_image_ids_.insert(image_id2);
  }

  return ba_success;
}

bool PtzIncrementalOptimizer::RegisterNextImage(long image_id)
{
  assert(NumRegImages() >= 2);
  assert(reg_image_ids_.find(image_id) == reg_image_ids_.end() && "Image cannot be registered multiple times");

  num_reg_trials_[image_id] += 1;

  for (const auto& match_info : matches_info_) {
    long i = match_info.src_img_idx;
    long j = match_info.dst_img_idx;
    cv::Mat H_j_i = match_info.H;
    if (H_j_i.empty())
      continue;

    if (reg_image_ids_.count(i) == 1 && j == image_id) {
      cameras_[j].K() = cameras_[i].K();
      cv::Mat R_j_i = cameras_[j].K().inv() * H_j_i * cameras_[i].K();
      cameras_[j].R() = R_j_i * cameras_[i].R();

      constexpr int max_iter = 100;
      constexpr double max_reproj_error = 100;
      KRTOptimizer optimizer(max_iter, max_reproj_error, KRTOptimizer::F);
      optimizer.SetInitParams(cameras_[j].K(), cameras_[j].R(), cameras_[j].t(), cameras_[j].dist());
      optimizer.Add2d2dConstraints(cameras_[i], features_[i].keypoints, features_[j].keypoints, match_info.matches);

      cv::Mat K, R, t, dist;
      bool ba_success = optimizer.Solve(K, R, t, dist);
      if (ba_success) {
        cameras_[j].K() = K;
        cameras_[j].R() = R;
        reg_image_ids_.insert(j);

        return true;
      }
      else {
        continue;
      }
    }
  }

  return false;
}

bool PtzIncrementalOptimizer::AdjustGlobalBundle()
{
  PLOGI << "Global bundle adjustment start";

  PTZRayOptimizer optimizer(features_, matches_info_, cameras_, reg_image_ids_, max_iter_, PTZRay);

  bool ba_success = optimizer.Solve(cameras_);
  double reprojection_error = optimizer.final_reproj_error_all();

  if (ba_success) {
    vector<long> reg_image_ids_sorted(reg_image_ids_.begin(), reg_image_ids_.end());
    sort(reg_image_ids_sorted.begin(), reg_image_ids_sorted.end());

    PLOGI << "Global bundle adjustment success. Rreprojecion error: " << reprojection_error;
  }
  else {
    PLOGI << "Global bundle adjustment failed. Rreprojecion error: " << reprojection_error;
  }

  return ba_success;
}

}  // namespace ptzcalib

File: python/ptzcalib/ptz_incremental_optimizer.py
"""Incremental PTZ calibration optimizer.

Python port of src/core/ptz_incremental_optimizer.{h,cc}.
Simplified implementation focusing on core functionality.
"""

from typing import Dict, List, Set, Optional
import numpy as np
import logging

from .types import Camera, ImageFeatures, MatchesInfo
from .ptzray_optimizer import PTZRayOptimizer, FactorType


class PtzIncrementalOptimizer:
    """Incremental bundle adjustment for PTZ camera calibration."""
    
    kMaxNumImages = 1000
    kBaGlobalImagesRatio = 1.1
    
    def __init__(self, features: List[ImageFeatures], matches_info: List[MatchesInfo],
                 cameras: List[Camera], max_iter: int, names: Optional[List[str]] = None):
        self.features_ = features
        self.matches_info_ = matches_info
        self.cameras_ = [cam.clone() for cam in cameras]
        self.max_iter_ = max_iter
        self.names_ = names if names else [f"img_{i}" for i in range(len(cameras))]
        
        self.init_image_pairs_: Set[int] = set()
        self.num_reg_trials_: Dict[int, int] = {}
        self.reg_image_ids_: Set[int] = set()
        self.seed_image_ids_: List[int] = []
    
    def set_seed_image_id(self, image_ids: List[int]):
        """Manually set seed image IDs for initialization."""
        self.seed_image_ids_ = list(image_ids)
    
    def solve(self, cameras_out: List[Camera]) -> tuple[bool, Set[int]]:
        """Run incremental optimization.
        
        Returns:
            (success, registered_image_ids)
        """
        if not self._check_valid():
            logging.error("Invalid input for PtzIncrementalOptimizer")
            return False, set()
        
        # Find and register initial image pair
        image_id1, image_id2 = -1, -1
        if not self._find_initial_image_pair(image_id1, image_id2):
            logging.error("Failed to find initial image pair")
            return False, set()
        
        if not self._register_initial_image_pair(image_id1, image_id2):
            logging.error(f"Failed to register initial pair ({image_id1}, {image_id2})")
            return False, set()
        
        logging.info(f"Successfully registered initial pair: {image_id1}, {image_id2}")
        
        # Incrementally register remaining images
        while self._num_reg_images() < len(self.cameras_):
            next_images = self._find_next_images()
            if not next_images:
                logging.info(f"No more images to register. Total: {self._num_reg_images()}/{len(self.cameras_)}")
                break
            
            registered_any = False
            for img_id in next_images:
                if img_id in self.reg_image_ids_:
                    continue
                if self._register_next_image(img_id):
                    logging.info(f"Registered image {img_id} ({self._num_reg_images()}/{len(self.cameras_)})")
                    registered_any = True
                    break
            
            if not registered_any:
                logging.warning("Failed to register any new image in this round")
                break
            
            # Run global BA periodically
            if self._num_reg_images() % max(1, int(self.kBaGlobalImagesRatio)) == 0:
                self._adjust_global_bundle()
        
        # Final global BA
        self._adjust_global_bundle()
        
        # Copy results
        for i in range(len(self.cameras_)):
            cameras_out[i] = self.cameras_[i].clone()
        
        return True, self.reg_image_ids_
    
    def _check_valid(self) -> bool:
        if not self.features_ or not self.cameras_:
            return False
        if len(self.features_) != len(self.cameras_):
            return False
        return True
    
    def _find_initial_image_pair(self, img_id1_out, img_id2_out) -> tuple[int, int]:
        """Find the best initial image pair."""
        if self.seed_image_ids_:
            first_candidates = self.seed_image_ids_
        else:
            first_candidates = self._find_first_initial_image()
        
        for img_id1 in first_candidates:
            second_candidates = self._find_second_initial_image(img_id1)
            for img_id2 in second_candidates:
                pair_id = self._image_pair_to_pair_id(img_id1, img_id2)
                if pair_id not in self.init_image_pairs_:
                    return img_id1, img_id2
        return -1, -1
    
    def _find_first_initial_image(self) -> List[int]:
        """Find candidate first seed images (high feature count, good matches)."""
        scores = []
        for i in range(len(self.features_)):
            num_features = len(self.features_[i].keypoints)
            num_matches = sum(1 for m in self.matches_info_ 
                            if (m.src_img_idx == i or m.dst_img_idx == i) and m.num_inliers > 0)
            scores.append((i, num_features * num_matches))
        scores.sort(key=lambda x: x[1], reverse=True)
        return [s[0] for s in scores[:10]]
    
    def _find_second_initial_image(self, image_id1: int) -> List[int]:
        """Find candidate second images for pairing with image_id1."""
        scores = []
        for m in self.matches_info_:
            if m.src_img_idx == image_id1:
                other_id = m.dst_img_idx
            elif m.dst_img_idx == image_id1:
                other_id = m.src_img_idx
            else:
                continue
            
            if m.num_inliers == 0:
                continue
            
            disparity = self._cal_pixel_diff(image_id1, other_id, m.matches)
            score = m.num_inliers * disparity
            scores.append((other_id, score))
        
        scores.sort(key=lambda x: x[1], reverse=True)
        return [s[0] for s in scores[:5]]
    
    def _find_next_images(self) -> List[int]:
        """Find next images to register (connected to registered images)."""
        candidates = []
        for m in self.matches_info_:
            if m.num_inliers == 0:
                continue
            
            if m.src_img_idx in self.reg_image_ids_ and m.dst_img_idx not in self.reg_image_ids_:
                candidates.append((m.dst_img_idx, m.num_inliers))
            elif m.dst_img_idx in self.reg_image_ids_ and m.src_img_idx not in self.reg_image_ids_:
                candidates.append((m.src_img_idx, m.num_inliers))
        
        candidates.sort(key=lambda x: x[1], reverse=True)
        return [c[0] for c in candidates[:10]]
    
    def _cal_pixel_diff(self, image_id1: int, image_id2: int, matches: List) -> float:
        """Calculate average pixel disparity between matched points."""
        if not matches:
            return 0.0
        
        total_diff = 0.0
        count = 0
        feat1 = self.features_[image_id1]
        feat2 = self.features_[image_id2]
        
        for dm in matches:
            if dm.queryIdx >= len(feat1.keypoints) or dm.trainIdx >= len(feat2.keypoints):
                continue
            pt1 = feat1.keypoints[dm.queryIdx].pt
            pt2 = feat2.keypoints[dm.trainIdx].pt
            diff = np.sqrt((pt1[0] - pt2[0])**2 + (pt1[1] - pt2[1])**2)
            total_diff += diff
            count += 1
        
        return total_diff / count if count > 0 else 0.0
    
    def _register_initial_image_pair(self, image_id1: int, image_id2: int) -> bool:
        """Register the initial image pair."""
        pair_id = self._image_pair_to_pair_id(image_id1, image_id2)
        self.init_image_pairs_.add(pair_id)
        
        # Set initial parameters (simplified: assume reasonable defaults)
        self._set_initial_image_pair_parameters(image_id1, image_id2)
        
        # Run PTZRay optimizer on this pair
        cam_ids = {image_id1, image_id2}
        optimizer = PTZRayOptimizer(
            self.features_, self.matches_info_, self.cameras_,
            cam_ids, self.max_iter_, FactorType.PTZRayDist
        )
        
        cameras_temp = [cam.clone() for cam in self.cameras_]
        success = optimizer.solve(cameras_temp)
        
        if success:
            self.cameras_[image_id1] = cameras_temp[image_id1].clone()
            self.cameras_[image_id2] = cameras_temp[image_id2].clone()
            self.reg_image_ids_.add(image_id1)
            self.reg_image_ids_.add(image_id2)
            return True
        
        return False
    
    def _register_next_image(self, image_id: int) -> bool:
        """Register a single new image."""
        self.num_reg_trials_[image_id] = self.num_reg_trials_.get(image_id, 0) + 1
        
        if self.num_reg_trials_[image_id] > 3:
            logging.warning(f"Image {image_id} failed registration {self.num_reg_trials_[image_id]} times, skipping")
            return False
        
        # Run PTZRay optimizer including this new image
        cam_ids = self.reg_image_ids_ | {image_id}
        optimizer = PTZRayOptimizer(
            self.features_, self.matches_info_, self.cameras_,
            cam_ids, self.max_iter_, FactorType.PTZRayDist
        )
        
        cameras_temp = [cam.clone() for cam in self.cameras_]
        success = optimizer.solve(cameras_temp)
        
        if success:
            for cid in cam_ids:
                self.cameras_[cid] = cameras_temp[cid].clone()
            self.reg_image_ids_.add(image_id)
            return True
        
        return False
    
    def _set_initial_image_pair_parameters(self, image_id1: int, image_id2: int):
        """Set initial camera parameters for the first pair (simplified)."""
        # In a full implementation, this would use essential matrix decomposition
        # Here we assume cameras are pre-initialized with reasonable values
        pass
    
    def _adjust_global_bundle(self) -> bool:
        """Run global bundle adjustment on all registered images."""
        if len(self.reg_image_ids_) < 2:
            return True
        
        optimizer = PTZRayOptimizer(
            self.features_, self.matches_info_, self.cameras_,
            self.reg_image_ids_, self.max_iter_, FactorType.PTZRayDist
        )
        
        cameras_temp = [cam.clone() for cam in self.cameras_]
        success = optimizer.solve(cameras_temp)
        
        if success:
            for cid in self.reg_image_ids_:
                self.cameras_[cid] = cameras_temp[cid].clone()
            logging.info(f"Global BA success with {len(self.reg_image_ids_)} images")
            return True
        
        logging.warning("Global BA failed")
        return False
    
    def _image_pair_to_pair_id(self, image_id1: int, image_id2: int) -> int:
        """Convert image pair to unique ID."""
        if image_id1 > image_id2:
            image_id1, image_id2 = image_id2, image_id1
        return image_id1 * self.kMaxNumImages + image_id2
    
    def _num_reg_images(self) -> int:
        return len(self.reg_image_ids_)
